pkgs/jni/android/src/main/java/com/github/dart_lang/jni/JniPlugin.java

```diff
@@ -7,48 +7,72 @@
 import android.app.Activity;
 import android.content.Context;
 import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
 import io.flutter.embedding.engine.plugins.FlutterPlugin;
 import io.flutter.embedding.engine.plugins.activity.ActivityAware;
 import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class JniPlugin implements FlutterPlugin, ActivityAware {
+  private static final ConcurrentHashMap<Long, JniPlugin> pluginMap = new ConcurrentHashMap<>();
+
+  private long engineId;
+  private volatile Context context;
+  private volatile Activity activity;
+
+  public static @NonNull Context getApplicationContext(long engineId) {
+    return Objects.requireNonNull(pluginMap.get(engineId)).context;
+  }
+
+  public static @Nullable Activity getActivity(long engineId) {
+    return Objects.requireNonNull(pluginMap.get(engineId)).activity;
+  }
 
   @Override
+  @SuppressWarnings("deprecation")
   public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
-    setup(binding.getApplicationContext());
+    //noinspection deprecation
+    engineId = binding.getFlutterEngine().getEngineId();
+    context = binding.getApplicationContext();
+    pluginMap.put(engineId, this);
   }
 
-  private void setup(Context context) {
-    initializeJni(context, getClass().getClassLoader());
+  @Override
+  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
+    context = null;
+    activity = null;
+    pluginMap.remove(engineId);
```

Comment:
How is the user notified that this `engineId` is no longer valid? Is there a possibility of race conditions where they call `getApplicationContext` etc after the `engineId` is removed from the map but before they've been notified? If so, those methods should probably fail gracefully (eg just return null if the `engineId` is invalid`).